File: app/src/main/java/org/autojs/autojs/network/entity/VersionInfo.java
package org.autojs.autojs.network.entity;

import static org.autojs.autojs.util.StringUtils.str;

import androidx.annotation.NonNull;

import org.autojs.autojs.util.UpdateUtils;
import org.autojs.autojs6.BuildConfig;
import org.autojs.autojs6.R;

import java.text.MessageFormat;

/**
 * Created by Stardust on 2017/9/20.
 * Modified by SuperMonster003 as of May 29, 2022.
 */
public class VersionInfo implements ExtendedVersionInfo {

    private String mVersionName;
    private int mVersionCode;

    private long mSize = -1;
    private String mDownloadUrl;
    private String mAbi;
    private String mFileName;

    public VersionInfo(@NonNull String propertiesFileRawString) {
        String regexVersionName = "VERSION_NAME=.+";
        String regexVersionCode = "VERSION_BUILD=.+";
        for (String string : propertiesFileRawString.split("\n")) {
            if (string.matches(regexVersionName)) {
                mVersionName = string.split("=")[1];
            } else if (string.matches(regexVersionCode)) {
                mVersionCode = Integer.parseInt(string.split("=")[1]);
            }
            if (mVersionName != null && mVersionCode > 0) {
                break;
            }
        }
    }

    public VersionInfo(@NonNull String versionName, int versionCode) {
        mVersionName = versionName;
        mVersionCode = versionCode;
    }

    public boolean isNewer() {
        return mVersionCode > BuildConfig.VERSION_CODE;
    }

    public boolean isNotIgnored() {
        return !UpdateUtils.isVersionIgnored(this);
    }

    @NonNull
    @Override
    public String toString() {
        if (mAbi == null) {
            return MessageFormat.format("{0}: {1} ({2})", str(R.string.text_version), mVersionName, mVersionCode);
        }
        return MessageFormat.format("{0}: {1} ({2}) [{3}]", str(R.string.text_version), mVersionName, mVersionCode, mAbi);
    }

    public String toSummary() {
        return MessageFormat.format("{0} ({1})", mVersionName, mVersionCode);
    }

    public static SimpleVersionInfo parseSummary(CharSequence summary) {
        int indexForVersionName = 0;
        int indexForVersionCode = 1;

        String versionName = null;
        int versionCode = -1;

        String[] split = summary.toString().split(" ");

        for (int i = 0; i < split.length; i++) {
            if (i == indexForVersionName) {
                versionName = split[i];
            } else if (i == indexForVersionCode) {
                versionCode = Integer.parseInt(split[i].replaceAll("[()]", ""));
            }
        }
        return new SimpleVersionInfo(versionName, versionCode);
    }

    public String getVersionName() {
        return mVersionName;
    }

    public int getVersionCode() {
        return mVersionCode;
    }

    public void setFileName(String fileName) {
        this.mFileName = fileName;
    }

    public void setSize(long size) {
        mSize = size;
    }

    public void setDownloadUrl(String downloadUrl) {
        mDownloadUrl = downloadUrl;
    }

    public String getFileName() {
        return mFileName;
    }

    public void setAbi(String abi) {
        mAbi = abi;
    }

    public long getSize() {
        return mSize;
    }

    public String getDownloadUrl() {
        return mDownloadUrl;
    }

    @Override
    public String getAbi() {
        return mAbi;
    }

}

File: app/src/main/java/org/autojs/autojs/network/entity/VersionInfo.kt
package org.autojs.autojs.network.entity

import android.util.Log
import org.autojs.autojs.util.StringUtils.str
import org.autojs.autojs.util.UpdateUtils.isVersionIgnored
import org.autojs.autojs6.BuildConfig
import org.autojs.autojs6.R
import kotlin.math.roundToInt

/**
 * Created by Stardust on Sep 20, 2017.
 * Modified by SuperMonster003 as of May 29, 2022.
 */
class VersionInfo : ExtendedVersionInfo {

    var versionName: String = ""
        private set

    var versionCode = 0
        private set

    val isNewer: Boolean
        get() = versionCode > BuildConfig.VERSION_CODE

    val isNotIgnored: Boolean
        get() = !isVersionIgnored(this)

    private var mFileName: String = ""
    private var mSize: Long = -1
    private var mDownloadUrl: String? = null
    private var mAbi: String? = null

    constructor(propertiesFileRawString: String) {
        val versionNameKey = str(R.string.property_key_app_version_name)
        val versionCodeKey = str(R.string.property_key_app_version_code)

        val regexVersionName = "$versionNameKey(?:</\\w+>)?=([\\w.]+)".toRegex()
        val regexVersionCode = "$versionCodeKey(?:</\\w+>)?=(\\d+)".toRegex()

        for (line in propertiesFileRawString.lineSequence().filter { it.isNotEmpty() }) {
            if (versionName.isBlank() && line.contains(regexVersionName)) {
                versionName = regexVersionName.find(line)?.groupValues?.get(1).orEmpty()
                Log.d(TAG, "versionName: $versionName")
            }
            if (versionCode <= 0 && line.contains(regexVersionCode)) {
                versionCode = regexVersionCode.find(line)?.groupValues?.get(1)?.toDouble()?.roundToInt() ?: -1
                Log.d(TAG, "versionCode: $versionCode")
            }
            if (versionName.isNotBlank() && versionCode > 0) {
                break
            }
        }
    }

    constructor(versionName: String, versionCode: Int) {
        this.versionName = versionName
        this.versionCode = versionCode
    }

    override fun toString() = when (mAbi) {
        null -> "${str(R.string.text_version)}: $versionName ($versionCode)"
        else -> "${str(R.string.text_version)}: $versionName ($versionCode) [$mAbi]"
    }

    fun toSummary() = "$versionName ($versionCode)"

    fun setFileName(fileName: String) {
        mFileName = fileName
    }

    fun setSize(size: Long) {
        mSize = size
    }

    fun setDownloadUrl(downloadUrl: String) {
        mDownloadUrl = downloadUrl
    }

    fun setAbi(abi: String) {
        mAbi = abi
    }

    override fun getFileName() = mFileName

    override fun getSize() = mSize

    override fun getDownloadUrl() = mDownloadUrl

    override fun getAbi() = mAbi

    companion object {

        fun parseSummary(summary: CharSequence): SimpleVersionInfo {
            val indexForVersionName = 0
            val indexForVersionCode = 1
            var versionName: String? = null
            var versionCode = -1
            val split = summary.toString().split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (i in split.indices) {
                if (i == indexForVersionName) {
                    versionName = split[i]
                } else if (i == indexForVersionCode) {
                    versionCode = split[i].replace("[()]".toRegex(), "").toInt()
                }
            }
            return SimpleVersionInfo(versionName, versionCode)
        }

        private val TAG = VersionInfo::class.java.simpleName

    }

}